concrete5-community-store/community_store
# Cart update across pages

### Description

This pull request is mainly about using JS broadcasting to make sure cart values are updated across open pages when updated on one page.

There are 3 spots where cart items and values need to be updated:

1. The JS cart from the utility links block
2. The /cart page
3. The /checkout page

The JS cart can be fully updated by JS and doesn't require a page reload. If the cart content is modified on another page, the cart is closed (if open) and the total is updated to show the new value.

The /cart page uses a single page with no element for the cart list so it's not possible to reliably update it with JS without a page reload. I would have to extract the cart itself in an element and since I'm sure lots of users are overriding that page to apply their own design, that would break lots of websites so not a good idea. So if the /cart page is open and the cart is updated elsewhere I simply reload the cart page. Not ideal but it works.

The /checkout page is a mixed bag. The cart itself is an element so I'm able to update it using JS, that's good. On the other hand, the discount part is not an element, it's part of the single page. So same reasoning here, I can't extract it as an element because that page is probably overridden by a lot of websites. As a result, if the cart content is modified elsewhere, the cart on the /checkout page is updated through JS without a page refresh. If a coupon is added or removed on another page, the /checkout page will reload.

### To test the cart update functionality:

1. Open a page with the cart block on it. Also, open the /cart and /checkout pages
2. Start by adding products to the cart and make sure the /cart page reloads and the /checkout page gets updated without reloading.
3. From the /cart page modify the cart content and make sure the JS cart on the other page the cart list on the /checkout page both get updated without page reload.
4. From the /cart page add a coupon and make sure the JS cart gets updated and the /checkout page reloads
5. From the /checkout page add or remove a coupon and make sure the JS cart gets updated and the /cart page reloads.

### Other modifications

This pull request also made sure the 3 third-party JS scripts used in CS were pulled in by npm:

1. sysend
2. chartist
3. chartist-plugin-tooltips

I then modified controller.php to pull the scripts from the new node_modules folder and removed the scripts from the JS and CSS folders.
I added a variable $npmPackages in the controller that lists version numbers for the packages. That makes it easier to update them so there's no need to modify it twice for each package every time.

I also fixed the JS call for chartist-plugin-tooltips. The symbol '$' was hard-coded instead of using the value from the settings on the store single page. I changed that.
